[13.0][MIG] sale_automatic_workflow_payment_mode: Migration to 13.0

victoralmau opened the pull request:
Migration to 13.0

Please @joao-p-marques and @pedrobaeza can you review it?

@Tecnativa TT31208

Commit:
[WIP][9.0] Add sale_automatic_workflow_payment_mode (#297)

[ADD] sale_automatic_workflow_payment_mode

pedrobaeza committed:
[MIG] Make modules uninstallable

[MIG] Rename manifest files

lmignon committed:
[MIG] sale_automatic_workflow_payment_mode: Upgrade to v10

Commit:
[IMP] sale_automatic_workflow_payment_mode: improve test coverage

lmignon committed:
[IMP] sale_automatic_workflow_payment_mode: OCA conventions

[IMP] sale_automatic_workflow_payment_mode: Use related fields for payment_filter_domain & char_domain widget

hugosantosred committed:
[MIG] sale_automatic_workflow_payment_mode to V11

murtuzasaleh committed:
[11.0] [MIG]sale_procurement_group_by_line: Migration to 11.0.

florian-dacosta committed:
Migration to version 12.0

sbidoul committed:
[IMP] sale_automatic_workflow_payment_mode cosmetics

liweijie0812 committed:
Translated using Weblate (Chinese (Simplified))

Currently translated at 20.0% (2 of 10 strings)

Translation: sale-workflow-12.0/sale-workflow-12.0-sale_automatic_workflow_payment_mode
Translate-URL: https://translation.odoo-community.org/projects/sale-workflow-12-0/sale-workflow-12-0-sale_automatic_workflow_payment_mode/zh_CN/

liweijie0812 committed:
Translated using Weblate (Chinese (Simplified))

Currently translated at 30.0% (3 of 10 strings)

Translation: sale-workflow-12.0/sale-workflow-12.0-sale_automatic_workflow_payment_mode
Translate-URL: https://translation.odoo-community.org/projects/sale-workflow-12-0/sale-workflow-12-0-sale_automatic_workflow_payment_mode/zh_CN/

liweijie0812 committed:
Translated using Weblate (Chinese (Simplified))

Currently translated at 40.0% (4 of 10 strings)

Translation: sale-workflow-12.0/sale-workflow-12.0-sale_automatic_workflow_payment_mode
Translate-URL: https://translation.odoo-community.org/projects/sale-workflow-12-0/sale-workflow-12-0-sale_automatic_workflow_payment_mode/zh_CN/

liweijie0812 committed:
Translated using Weblate (Chinese (Simplified))

Currently translated at 100.0% (10 of 10 strings)

Translation: sale-workflow-12.0/sale-workflow-12.0-sale_automatic_workflow_payment_mode
Translate-URL: https://translation.odoo-community.org/projects/sale-workflow-12-0/sale-workflow-12-0-sale_automatic_workflow_payment_mode/zh_CN/

Galoperd committed:
Translated using Weblate (Spanish)

Currently translated at 90.0% (9 of 10 strings)

Translation: sale-workflow-12.0/sale-workflow-12.0-sale_automatic_workflow_payment_mode
Translate-URL: https://translation.odoo-community.org/projects/sale-workflow-12-0/sale-workflow-12-0-sale_automatic_workflow_payment_mode/es/

Commit:
Add sale_automatic_workflow_job

Use Queue Jobs to process the Sales Automatic Workflow actions.

The default behavior of the automatic workflow module is to use a
scheduled action that searches all the record that need a workflow
action and sequentially process all of them.

It can hit some limits when the number of records is too high.

This module keeps the scheduled action to search the records, but
instead of directly executing the actions (confirm a sales order,
create invoices for a sales order, validate invoices, ...), it
creates one job per operation to do.

It uses an identity key on the jobs so it will not create the same
job for the same record and same operation twice.

~

I needed to extract methods in `sale_automatic_workflow` in order
to be able to execute them as jobs.

A new decorator "job_auto_delay" (to eventually move in queue_job) makes
these methods automatically delayed when called.

sysadminmatmoz committed:
Translated using Weblate (Slovenian)

Currently translated at 30.0% (3 of 10 strings)

Translation: sale-workflow-12.0/sale-workflow-12.0-sale_automatic_workflow_payment_mode
Translate-URL: https://translation.odoo-community.org/projects/sale-workflow-12-0/sale-workflow-12-0-sale_automatic_workflow_payment_mode/sl/

renatonlima committed:
[FIX] pep8 and pylint

britoederr committed:
Translated using Weblate (Portuguese (Brazil))

Currently translated at 100.0% (10 of 10 strings)

Translation: sale-workflow-12.0/sale-workflow-12.0-sale_automatic_workflow_payment_mode
Translate-URL: https://translation.odoo-community.org/projects/sale-workflow-12-0/sale-workflow-12-0-sale_automatic_workflow_payment_mode/pt_BR/

danimv5 committed:
Translated using Weblate (Catalan)

Currently translated at 100.0% (10 of 10 strings)

Translation: sale-workflow-12.0/sale-workflow-12.0-sale_automatic_workflow_payment_mode
Translate-URL: https://translation.odoo-community.org/projects/sale-workflow-12-0/sale-workflow-12-0-sale_automatic_workflow_payment_mode/ca/

danimv5 committed:
Translated using Weblate (Spanish)

Currently translated at 100.0% (10 of 10 strings)

Translation: sale-workflow-12.0/sale-workflow-12.0-sale_automatic_workflow_payment_mode
Translate-URL: https://translation.odoo-community.org/projects/sale-workflow-12-0/sale-workflow-12-0-sale_automatic_workflow_payment_mode/es/

victoralmau committed:
[IMP] sale_automatic_workflow_payment_mode: black, isort, prettier

victoralmau committed:
[MIG] sale_automatic_workflow_payment_mode: Migration to 13.0

pedrobaeza commented:
There are already 3 opened PRs for this module:

#1283
#1183
#1080 

But checking them, they seems not finished, so closing them.

pedrobaeza commented:
Please check the reviews in #1283

bizzappdev commented:
> Please check the reviews in #1283

Not sure what is the reason, But I have already solved those review changes in #1633

pedrobaeza commented:
@bizzappdev as you don't follow the proper title scheme for the PR, they are not findable through general search. Please follow such scheme in all your PRs, and you will avoid surprises, but I'm OK to switch to review that one, giving it priority and see if we finally get this merged.

bizzappdev commented:
@pedrobaeza  Extremely sorry for now the following standard.

victoralmau commented:
Sorry, can you check it to mergue and unblock other PRs that depend on this one?

bizzappdev commented:
> Sorry, can you check it to mergue and unblock other PRs that depend on this one?

sorry. Do I have to do anything?

joao-p-marques commented on sale_automatic_workflow_payment_mode/readme/CONTRIBUTORS.rst:
```diff
@@ -0,0 +1,2 @@
+* Guewen Baconnier <guewen.baconnier@camptocamp.com>
+* Sodexis <dev@sodexis.com>
```
You can add Tecnativa here with your name :)

Also, you can re-generate the README if you change it.

joao-p-marques reviewed (approved):
Code review :+1: 

Just a small suggestion, but non-blocking.

pedrobaeza commented on sale_automatic_workflow_payment_mode/data/automatic_workflow_data.xml:
```diff
@@ -7,8 +7,10 @@
         <field name="name">Automatic Workflow Payment Filter</field>
-        <field name="model_id">account.invoice</field>
-        <field name="domain">[('state', '=', 'open')]</field>
+        <field name="model_id">account.move</field>
```
This field is a many2one:

```suggestion
<field name="model_id" ref="account.model_account_move" />
```

victoralmau replied:
I am checking v13 and v14 in runbot and the `model_id` field in `ir.filters` model is a selection field.

pedrobaeza replied:
Oh, sorry, I have mixed it with `ir.rule`.  The name of the field doesn't also help too much.

pedrobaeza commented on sale_automatic_workflow_payment_mode/data/automatic_workflow_data.xml:
```diff
@@ -7,8 +7,10 @@
         <field name="name">Automatic Workflow Payment Filter</field>
-        <field name="model_id">account.invoice</field>
-        <field name="domain">[('state', '=', 'open')]</field>
+        <field name="model_id">account.move</field>
```
And this requires migration script.

pedrobaeza reviewed (changes requested):
There are other references to `account.invoice`, which disappears in this version.

victoralmau commented:
> There are other references to `account.invoice`, which disappears in this version.

Thanks, changes done.

pedrobaeza commented:
I still see references to `account.invoice` model.

pedrobaeza commented on sale_automatic_workflow_payment_mode/models/automatic_workflow_job.py:
```diff
@@ -0,0 +1,64 @@
+    def run_with_workflow(self, sale_wkf):
+        workflow_domain = [("workflow_process_id", "=", sale_wkf.id)]
+        res = super(AutomaticWorkflowJob, self).run_with_workflow(sale_wkf)
+        if sale_wkf.register_payment:
```
I don't get how this is related with `payment_mode`. This is something standard...

victoralmau replied:
IMO, it is not necessary to add anything extra here since for each payment mode a `sale.workflow.process` is defined, and therefore in the onchange of the `payment_mode_id` of the order, you have already defined the correct workflow; therefore, when this action is executed, it is no longer necessary to take into account the payment mode.

pedrobaeza replied:
The question is that this is registering a payment, which is something standard, not linked with `payment.mode`. Why is this code here? I understand this should go in the main module if needed, and if it's modifying a bit such registering, then it should be done by a hook on main module + the extension method here.

victoralmau replied:
Ok, I understand.
I have added the `_register_payments()` function to the `sale_automatic_workflow` and also, in `sale_automatic_workflow_payment_mode` I have added only the necessary related to the payment mode.

pedrobaeza commented on sale_automatic_workflow/models/automatic_workflow_job.py:
```diff
@@ -140,6 +140,35 @@ def _sale_done(self, sale_done_filter):
+            "invoice_ids": [(6, 0, invoice.ids)],
+            "amount": invoice.residual,
+            "payment_date": fields.Date.context_today(self),
+            "communication": invoice.reference or invoice.number,
```
`invoice.number` doesn't exist anymore. It should be `invoice.name` for customer invoices, and `invoice.reference` for vendor bills.

pedrobaeza commented on sale_automatic_workflow_payment_mode/models/automatic_workflow_job.py:
```diff
@@ -24,41 +22,22 @@ def run_with_workflow(self, sale_wkf):
+                invoice.id,
+                invoice.payment_mode_id.id,
+            )
+        else:
```
No need of the else, and put `return super...`

pedrobaeza commented on sale_automatic_workflow_payment_mode/views/sale_workflow_process_view.xml:
```diff
@@ -25,15 +25,15 @@ The licence is in the file __openerp__.py -->
                         name="payment_filter_domain"
```
This should go also to the main module.

victoralmau replied:
Ok, some fields have been moved to the main addon `sale_automatic_workflow`.

pedrobaeza reviewed (changes requested):
You have made a mix on the 2 commits, as you removing the code for registering payments in the migration commit. Better to do everything in the same migration commit, stating on the commit message what's going on.

victoralmau committed:
[IMP] sale_automatic_workflow:
- Move register_payment option from sale_automatic_workflow_payment_mode to
sale_automatic_workflow.
- Add _register_payments function in sale_automatic_workflow.
TT31208

pedrobaeza commented on sale_automatic_workflow/models/sale_workflow_process.py:
```diff
@@ -115,3 +115,17 @@ def _default_filter(self, xmlid):
         ),
     )
+
+    @api.model
```
Don't put this method in the middle of the field definition.

pedrobaeza commented:
/ocabot merge minor
